proposals/NNNN-specialize.md

```diff
@@ -0,0 +1,193 @@
+
+In some cases, though, this information is obscured from the compiler. This proposal introduces a new attribute, `@specialize`, which allows the author of a generic function to generate pre-specialized versions of that function for specific types. When the unspecialized version of the function is called with one of those types, the compiler will generate code that will re-dispatch to those prespecialized versions if available.
+
+## Motivation
+
+Consider the following generic function that sums an array of any binary integer type:
+
+```swift
+extension Sequence where Element: BinaryInteger {
+  func sum() -> Double {
+    reduce(0) { $0 + Double($1) }
+  }
+}
+```
+
+If you call this function directly on an array of a specific integer type (e.g. `Array<Int>`):
+
+```
+let arrayOfInt: [Int] = // ...
+let result = arrayOfInt.sum()
+```
+
+in optimized builds, the Swift compiler it will generate a specialized version of `sum` for that type. 
+
+If you inspect the binary, you will see this specialized version under a symbol `_$sST3sumSz7ElementRpzrlEAASdyFSaySiG_Tg5`, which demangles to `generic specialization <[Swift.Int]> of (extension in example):Swift.Sequence< where A.Element: Swift.BinaryInteger>.sum() -> Swift.Double`, alongside the unspecialized version.
+
+This specialized version of `sum` will be optimized specifically for `Array<Int>`. It can move a pointer directly over the the array's buffer, loading the elements into registers directly from memory. On modern hardware, it can make use of dedicated instructions to convert the them to floating point. It can do this because it can see the implementation of `sum`, and can see the exact types on which it is being called. What exact assembly instructions are generated will differ significantly between e.g. `Array<Int8>.sum` and `Array<UInt64>.sum`.
+
+Here is that `Array<Int>`-specialized code when compiled to x86-64 with `swiftc -Osize`:
+
+```asm
+_$sST3sumSz7ElementRpzrlEAASdyFSaySiG_Tg5:
+        mov     rax, qword ptr [rdi + 16]
+        xorpd   xmm0, xmm0
+        test    rax, rax
+        je      .LBB1_3
+        xor     ecx, ecx
+.LBB1_2:
+        cvtsi2sd        xmm1, qword ptr [rdi + 8*rcx + 32]
+        inc     rcx
+        addsd   xmm0, xmm1
+        cmp     rax, rcx
+        jne     .LBB1_2
+.LBB1_3:
+        ret
+```
+
+Now consider some code that places an optimization barrier between the call site and the concrete type:
+
+```swift
+protocol Summable: Sequence where Element: BinaryInteger { }
+extension Array: Summable where Element: BinaryInteger { }
+
+var summable: any Summable = []
```

Comment:
-> `var summable: any Summable = Array<Int>()` to avoid the `error: type 'Any' cannot conform to 'BinaryInteger'`

Reply:
```suggestion
var summable: any Summable
```